apps/go-test-caching/src/pipeline.ts

```diff
@@ -0,0 +1,304 @@
+import * as core from "@actions/core";
+import * as github from "@actions/github";
+
+import { Inputs } from "./main.js";
+import { logSection, logObject } from "./log.js";
+import { getHashFile, commitTestHashIndex } from "./github.js";
+import { listPackages } from "./pipeline/filter.js";
+import {
+  compileConcurrent,
+  validateCompilationResultsOrThrow,
+} from "./pipeline/build.js";
+import { comparePackagesToIndex, hashFile } from "./pipeline/hash.js";
+import { runConcurrent, validateRunResultsOrThrow } from "./pipeline/run.js";
+
+/**
+ * General information from the Execa call.
+ */
+export interface ExecInfo {
+  command: string;
+  exitCode: number;
+  cwd: string;
+  durationMs: number;
+}
+
+/**
+ * A Go package with an import path and directory.
+ * The import path being the package's name, and the directory being the path to the package within the repository.
+ *
+ * The directory is important because when go executes tests, the CWD is typically the package's directory.
+ * When running tests binaries separately, we need to properly set the CWD to the package's directory.
+ */
+export type GoPackage = { importPath: string; directory: string };
+
+/**
+ * The result of filtering which packages are applicable to this event.
+ *
+ * This is the first step in the processing pipeline.
+ */
+export interface LocalPackages {
+  [packageName: string]: GoPackage;
+}
+
+/**
+ * STEP 1: Filter Tests
+ */
+export async function getTestPackages(inputs: Inputs): Promise<LocalPackages> {
+  logSection("Locating Packages");
+  // if (inputs.tagFilter) {
+  //   return findTaggedTestPackages(inputs.moduleDirectory, inputs.tagFilter);
+  // }
+
+  return listPackages(inputs.moduleDirectory);
+}
+
+/**
+ * The result of compiling the filtered packages.
+ * Some of these packages contain no tests, and therefore no test binary is outputted.
+ * These packages should be filtered out before the next step.
+ *
+ * This is the second step in the processing pipeline.
+ */
+interface CompileInfo {
+  binary: string;
+  log: string;
+  execution: ExecInfo;
+}
+export interface CompiledPackages {
+  [packageName: string]: LocalPackages[string] & {
+    compile: CompileInfo;
+  };
+}
+
+/**
+ * STEP 2: Build Test Binaries
+ */
+export async function buildTestBinaries(
+  inputs: Inputs,
+  packages: LocalPackages,
+): Promise<CompiledPackages> {
+  logSection("Build Tests");
+
+  const maxBuildConcurrency = parseInt(core.getInput("build-concurrency")) || 4;
```

Comment:
The yaml default is `8` but here we have `4`. Should they be aligned? Could they be dynamic like GOMAXPROCS rather than hard-coded?

Reply:
Defaulted both run/build concurrency to number of CPU cores.

Reply:
5c582558ed60cee185a7f1b8c70c29b829263a72